File: embedding-worker/handler.py
import json
import os
import logging
import boto3
import psycopg2
import psycopg2.extras
from concurrent.futures import ThreadPoolExecutor, as_completed

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger(__name__)

SUPABASE_DB_URL = os.environ["SUPABASE_DB_URL"]
BEDROCK_MODEL_ID = os.environ.get("BEDROCK_MODEL_ID", "amazon.titan-embed-text-v2:0")
EMBEDDING_DIMENSIONS = int(os.environ.get("EMBEDDING_DIMENSIONS", "1024"))
MAX_WORKERS = int(os.environ.get("MAX_WORKERS", "10"))
AWS_REGION = os.environ.get("AWS_REGION", "eu-west-2")

s3 = boto3.client("s3")
bedrock = boto3.client("bedrock-runtime", region_name=AWS_REGION)


def lambda_handler(event, context):
    for record in event.get("Records", []):
        # SQS wraps the S3 event in the message body
        s3_event = json.loads(record["body"])

        for s3_record in s3_event.get("Records", []):
            bucket = s3_record["s3"]["bucket"]["name"]
            key = s3_record["s3"]["object"]["key"]

            log.info("Processing chunks file: s3://%s/%s", bucket, key)
            process_chunks_file(bucket, key)


def process_chunks_file(bucket: str, key: str):
    # Download and parse chunks.json
    response = s3.get_object(Bucket=bucket, Key=key)
    payload = json.loads(response["body"].read())

    chunks = payload.get("chunks", [])
    if not chunks:
        log.warning("No chunks found in %s", key)
        return

    log.info("Embedding %d chunks for document %s", len(chunks), payload["document_id"])

    # Embed all chunks concurrently
    embedded = embed_chunks(chunks)

    # Bulk insert into Supabase
    insert_chunks(embedded, payload)

    log.info("Indexed %d chunks for document %s", len(embedded), payload["document_id"])


def embed_chunks(chunks: list) -> list:
    results = [None] * len(chunks)

    def embed_one(index, chunk):
        embedding = get_embedding(chunk["text"])
        return index, {**chunk, "embedding": embedding}

    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
        futures = {
            executor.submit(embed_one, i, chunk): i
            for i, chunk in enumerate(chunks)
        }
        for future in as_completed(futures):
            index, result = future.result()
            results[index] = result

    return results


def get_embedding(text: str) -> list[float]:
    body = json.dumps({
        "inputText": text,
        "dimensions": EMBEDDING_DIMENSIONS,
        "normalize": True,
    })

    response = bedrock.invoke_model(
        modelId=BEDROCK_MODEL_ID,
        body=body,
        contentType="application/json",
        accept="application/json",
    )

    result = json.loads(response["body"].read())
    return result["embedding"]


def insert_chunks(chunks: list, payload: dict):
    rows = [
        (
            chunk["chunk_id"],
            chunk["document_id"],
            chunk["text"],
            chunk["embedding"],
            chunk.get("chunk_type"),
            chunk.get("heading_path", []),
            chunk.get("parent_heading"),
            chunk.get("sequence"),
            chunk.get("page_start"),
            chunk.get("page_end"),
            chunk.get("estimated_token_count"),
            payload.get("source_bucket"),
            payload.get("source_key"),
            payload.get("processed_at"),
        )
        for chunk in chunks
    ]

    sql = """
        INSERT INTO document_chunks (
            chunk_id, document_id, text, embedding,
            chunk_type, heading_path, parent_heading, sequence,
            page_start, page_end, estimated_token_count,
            source_bucket, source_key, processed_at
        )
        VALUES %s
        ON CONFLICT (chunk_id) DO UPDATE SET
            embedding             = EXCLUDED.embedding,
            text                  = EXCLUDED.text,
            chunk_type            = EXCLUDED.chunk_type,
            heading_path          = EXCLUDED.heading_path,
            parent_heading        = EXCLUDED.parent_heading,
            sequence              = EXCLUDED.sequence,
            page_start            = EXCLUDED.page_start,
            page_end              = EXCLUDED.page_end,
            estimated_token_count = EXCLUDED.estimated_token_count,
            processed_at          = EXCLUDED.processed_at
    """

    with psycopg2.connect(SUPABASE_DB_URL) as conn:
        with conn.cursor() as cur:
            psycopg2.extras.execute_values(cur, sql, rows)
        conn.commit()

File: workers/embedding-worker/handler.js
const { S3Client, GetObjectCommand } = require("@aws-sdk/client-s3");
const {
  BedrockRuntimeClient,
  InvokeModelCommand,
} = require("@aws-sdk/client-bedrock-runtime");
const { Pool } = require("pg");

const s3 = new S3Client({ region: process.env.AWS_REGION ?? "eu-west-2" });

const bedrock = new BedrockRuntimeClient({
  region: process.env.AWS_REGION ?? "eu-west-2",
});

let pool = null;
function getPool() {
  if (!pool) pool = new Pool({ connectionString: process.env.SUPABASE_DB_URL });
  return pool;
}

exports.handler = async (event) => {
  for (const sqsRecord of event.Records) {
    const s3Event = JSON.parse(sqsRecord.body);

    for (const s3Record of s3Event.Records ?? []) {
      const bucket = s3Record.s3.bucket.name;
      const key = decodeURIComponent(
        s3Record.s3.object.key.replace(/\+/g, " "),
      );
      console.log(`Processing: s3://${bucket}/${key}`);
      await processChunksFile(bucket, key);
    }
  }
};

async function processChunksFile(bucket, key) {
  const response = await s3.send(
    new GetObjectCommand({ Bucket: bucket, Key: key }),
  );

  const raw = await response.Body?.transformToString();
  if (!raw) throw new Error(`Empty S3 response: s3://${bucket}/${key}`);

  const payload = JSON.parse(raw);
  const { chunks } = payload;

  if (!chunks?.length) {
    console.warn(`No chunks found in ${key}`);
    return;
  }

  console.log(
    `Embedding ${chunks.length} chunks for document ${payload.document_id}`,
  );

  const embedded = await embedConcurrently(chunks);
  await upsertChunks(embedded, payload);

  console.log(
    `Indexed ${embedded.length} chunks for document ${payload.document_id}`,
  );
}

async function embedConcurrently(chunks) {
  const concurrency = parseInt(process.env.CONCURRENCY ?? "10");
  const results = [];

  for (let i = 0; i < chunks.length; i += concurrency) {
    const batch = chunks.slice(i, i + concurrency);
    const batchResults = await Promise.all(
      batch.map(async (chunk) => ({
        ...chunk,
        embedding: await getEmbedding(chunk.text),
      })),
    );
    results.push(...batchResults);
  }

  return results;
}

async function getEmbedding(text) {
  const command = new InvokeModelCommand({
    modelId: process.env.BEDROCK_MODEL_ID ?? "amazon.titan-embed-text-v2:0",
    body: JSON.stringify({
      inputText: text,
      dimensions: parseInt(process.env.EMBEDDING_DIMENSIONS ?? "1024"),
      normalize: true,
    }),
    contentType: "application/json",
    accept: "application/json",
  });

  const response = await bedrock.send(command);
  const result = JSON.parse(new TextDecoder().decode(response.body));
  return result.embedding;
}

async function upsertChunks(chunks, payload) {
  if (!chunks.length) return;

  const db = getPool();

  const COLUMNS = [
    "chunk_id",
    "document_id",
    "text",
    "embedding",
    "chunk_type",
    "heading_path",
    "parent_heading",
    "sequence",
    "page_start",
    "page_end",
    "estimated_token_count",
    "source_bucket",
    "source_key",
    "processed_at",
  ];

  const values = [];

  const placeholders = chunks.map((chunk, i) => {
    const offset = i * COLUMNS.length;
    values.push(
      chunk.chunk_id,
      chunk.document_id,
      chunk.text,
      `[${chunk.embedding.join(",")}]`,
      chunk.chunk_type ?? null,
      chunk.heading_path ?? [],
      chunk.parent_heading ?? null,
      chunk.sequence ?? null,
      chunk.page_start ?? null,
      chunk.page_end ?? null,
      chunk.estimated_token_count ?? null,
      payload.source_bucket,
      payload.source_key,
      payload.processed_at,
    );
    return `(${COLUMNS.map((_, j) => `$${offset + j + 1}`).join(", ")})`;
  });

  const sql = `
    INSERT INTO document_chunks (${COLUMNS.join(", ")})
    VALUES ${placeholders.join(", ")}
    ON CONFLICT (chunk_id) DO UPDATE SET
      embedding             = EXCLUDED.embedding,
      text                  = EXCLUDED.text,
      chunk_type            = EXCLUDED.chunk_type,
      heading_path          = EXCLUDED.heading_path,
      parent_heading        = EXCLUDED.parent_heading,
      sequence              = EXCLUDED.sequence,
      page_start            = EXCLUDED.page_start,
      page_end              = EXCLUDED.page_end,
      estimated_token_count = EXCLUDED.estimated_token_count,
      processed_at          = EXCLUDED.processed_at
  `;

  await db.query(sql, values);
}
